### ISSUE TITLE
chemspiderapi

### ISSUE THREAD

**Issue by RaoulWolf** [ropensci/software-review#329]
Submitting Author Name: Raoul Wolf
Submitting Author Github Handle: @RaoulWolf
Repository:  https://github.com/RaoulWolf/chemspiderapi
Version submitted: 0.0.2  2021-03-15
Editor: @jooolia
Reviewers: @rajarshi, @yufree, @data-datum
Due date for @rajarshi: 2021-03-15
Due date for @yufree: 2021-03-15
Due date for @data-datum: 2021-03-15
Archive: TBD  
Version accepted: TBD   

---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: chemspiderapi
Type: Package
Title: R Wrapper for ChemSpider's API Services
Version: 0.0.2
Authors@R: person("Raoul", "Wolf", email = "raoul.wolf@niva.no", role = c("aut", "cre"))
Description: ChemSpider has announced a fundamental change to the syntax of their API services in late 2018.
    This package provides convenience wrappers for the new API functionalities of ChemSpider, as well as complementary functions.
License: MIT + file LICENSE
Depends: R (>= 3.5.0)
Imports: 
    curl, 
    jsonlite
Suggests: 
    covr,
    keyring,
    knitr,
    magick,
    memoise,
    ratelimitr,
    rmarkdown,
    testthat
URL: https://github.com/RaoulWolf/chemspiderapi
Encoding: UTF-8
LazyData: true
ByteCompile: true
RoxygenNote: 6.1.1
VignetteBuilder: knitr
```

## Scope 

- Please indicate which category or categories from our [package fit policies](https://ropensci.github.io/dev_guide/policies.html#package-categories) this package falls under: (Please check an appropriate box below. If you are unsure, we suggest you make a pre-submission inquiry.):

	- [x] data retrieval
	- [ ] data extraction
	- [ ] database access
	- [ ] data munging
	- [ ] data deposition
	- [ ] reproducibility
	- [ ] geospatial data
	- [ ] text analysis
	

- Explain how and why the package falls under these categories (briefly, 1-2 sentences):

The `chemspiderapi` package is an easy-to-use R interface to use all new ChemSpider API functionalities, as introduced in ChemSpiders complete redesign of its API structure late 2018.

-   Who is the target audience and what are scientific applications of this package?  

Researchers and citizen scientists who work on anything chemistry-related and need to routinely query against any of ChemSpider's API services.

-   Are there other R packages that accomplish the same thing? If so, how does yours differ or meet [our criteria for best-in-category](https://ropensci.github.io/dev_guide/policies.html#overlap)?

The rOpenSci-maintained package webchem aims at offering (currently outdated / not working) functionality for accessing ChemSpider's API services. However, the redesign of ChemSpider's API structures in late 2018 broke all available functionalities.

-   If you made a pre-submission enquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

Pre-submission enquiry: https://github.com/ropensci/software-review/issues/294
@melvidoni 

## Technical checks

Confirm each of the following by checking the box.  This package:

- [x] does not violate the Terms of Service of any service it interacts with. 
- [x] has a CRAN and OSI accepted license.
- [x] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme). 
- [x] includes [documentation with examples for all functions](https://ropensci.github.io/dev_guide/building.html#documentation).
- [x] contains a vignette with examples of its essential functions and uses.
- [x] has a [test suite](https://ropensci.github.io/dev_guide/building.html#testing).
- [x] has [continuous integration](https://ropensci.github.io/dev_guide/ci.html), including reporting of test coverage using services such as Travis CI, Coveralls and/or CodeCov.

## Publication options

- [x] Do you intend for this package to go on CRAN?  
- [x] Do you wish to automatically submit to the [Journal of Open Source Software](http://joss.theoj.org/)? If so:

<details>
 <summary>JOSS Options</summary>  

  - [x] The package has an **obvious research application** according to [JOSS's definition](https://joss.readthedocs.io/en/latest/submitting.html#submission-requirements).
    - [x] The package contains a `paper.md` matching [JOSS's requirements](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain) with a high-level description in the package root or in `inst/`.
    - [ ] The package is deposited in a long-term repository with the DOI: 
    - (*Do not submit your package separately to JOSS*)  
  
</details>

- [ ] Do you wish to submit an Applications Article about your package to [Methods in Ecology and Evolution](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/)? If so:  

<details>
<summary>MEE Options</summary>  

- [ ] The package is novel and will be of interest to the broad readership of the journal. 
- [ ] The manuscript describing the package is no longer than 3000 words.
- [ ] You intend to archive the code for the package in a long-term repository which meets the requirements of the journal (see [MEE's Policy on Publishing Code](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/journal-resources/policy-on-publishing-code.html))
- (*Scope: Do consider MEE's [Aims and Scope](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/aims-and-scope/read-full-aims-and-scope.html) for your manuscript. We make no guarantee that your manuscript will be within MEE scope.*)
- (*Although not required, we strongly recommend having a full manuscript prepared when you submit here.*)
- (*Please do not submit your package separately to Methods in Ecology and Evolution*)

</details>

## Code of conduct

- [x] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.github.io/dev_guide/policies.html#code-of-conduct) during the review process and in maintaining my package should it be accepted.

**Comment by annakrystalli** [comment]
Hello @RaoulWolf, just in the process of assigning an editor. In the meantime, just checking, did you mean to submit your fork of the repository?

**Comment by RaoulWolf** [comment]
Hi @annakrystalli, yes I submitted my fork on purpose. I couldn't get an OK to incorporate AppVeyor, CodeCov and Travis CI on the other repository, so I forked it into my own account and added the functionality.

**Comment by annakrystalli** [comment]
OK, that's fine. I was just wondering whether further development would be done in the upstream fork or yours?

Also, we require test coverage of at least 75% before review. Your codecov badge indicates only 10% test coverage. Is there a reason for that?

**Comment by RaoulWolf** [comment]
Further development would be at the upstream fork, yes. It looks like I might be able to get Circle CI and CodeCov up and running on the upstream fork. Would that be preferable then?

Coverage is an issue indeed, but comes with the nature of the package. I've increased the coverage considerably (>30%), but I'm not sure how to increase it further. The main issue is that I do not want the tests to run the actual API queries because there's a rate limit on the queries. Any recommendations are more than welcome :)

**Comment by annakrystalli** [comment]
Hey @RaoulWolf.

Thanks for the updates and your efforts with improving test coverage!

 Regarding testing the API, have you had a look at package [httptest](https://cran.r-project.org/web/packages/httptest/index.html)?  This should help with what you're trying to achieve.

**Comment by RaoulWolf** [comment]
Hey @annakrystalli,

thanks for the heads up! I'll give `httptest` a try asap :)

**Comment by RaoulWolf** [comment]
Hey @annakrystalli,

thanks for the heads up! I'll give `httptest` a try asap :)

**Comment by sckott** [comment]
@RaoulWolf  Another option for caching tests (not doing real requests) is the vcr package - but only will work if you use crul or httr instead of curl

**Comment by RaoulWolf** [comment]
Thanks @sckott for the heads up on `vcr`. 

As you indicated, `vcr` (currently?) only works for `crul` or `httr`; also `httptest` does not work for `curl`. The reason I chose `curl` for the package was the relative flexibility over `httr` when assembling headers and data fields. In some instances I was not able to set up working API calls using `httr`. 

I will take a further look into `crul` - of course I would prefer a solution where `httptest` or `vcr` would offer support for `curl` 😃 

In the meantime, the upstream fork at NIVANorge (https://github.com/NIVANorge/chemspiderapi) has support for AppVeyor, CircleCi, Travis CI, and CodeCov. I assume it's fair game to change the repo address in the OP?

**Comment by sckott** [comment]
yeah `curl` is a great choice. jeroen is considering integration for webmockr (and therefore vcr) https://github.com/jeroen/curl/pull/174 but not sure if it will happen

**Comment by RaoulWolf** [comment]
Unfortunately I did not get `crul` to work for all possible functionalities; I have thus decided to stick with `curl`. 

Otherwise I was able to bump up the test coverage to over 40% without using the actual API query functionalities 👍 I'm still eager to improve coverage, but I'm not sure how to at this moment...? Any help/recommendation is highly appreciated!

I will also update the OP to link to the upstream repository (https://github.com/NIVANorge/chemspiderapi).

**Comment by noamross** [comment]
There's also https://github.com/nealrichardson/httptest

**Comment by maelle** [comment]
:wave: @RaoulWolf! any update?

**Comment by maelle** [comment]
:wave: @RaoulWolf! any update?

**Comment by maelle** [comment]
@RaoulWolf, you can ask any question reg testing on https://discuss.ropensci.org :-)

**Comment by noamross** [comment]
⚠️⚠️⚠️⚠️⚠️

***In the interest of reducing load on reviewers and editors as we manage the COVID-19 crisis, rOpenSci is temporarily pausing new submissions for software peer review for 30 days (and possibly longer). Please check back here again after 17 April for updates.***

***In this period new submissions will not be handled, nor new reviewers assigned. Reviews and responses to reviews will be handled on a 'best effort' basis, but no follow-up reminders will be sent.***

***Other rOpenSci community activities continue. We express our continued great appreciation for the work of our authors and reviewers. Stay healthy and take care of one other.***

***The rOpenSci Editorial Board***

⚠️⚠️⚠️⚠️⚠️

**Comment by stitam** [comment]
Thank you @RaoulWolf for your work on chemspiderapi! I am one of the maintainers of webchem. I would like to note that since 2019 September webchem can also access the new ChemSpider web service (https://github.com/ropensci/webchem/issues/149), so I think there might be a package overlap issue (https://devguide.ropensci.org/policies.html#overlap). We would be very happy if this could be resolved, please contact us if you wish to dicuss it in more detail.

**Comment by annakrystalli** [comment]
⚠️⚠️⚠️⚠️⚠️
In the interest of reducing load on reviewers and editors as we manage the COVID-19 crisis, rOpenSci new submissions for software peer review are paused. 

In this period new submissions will not be handled, nor new reviewers assigned. Reviews and responses to reviews will be handled on a 'best effort' basis, but no follow-up reminders will be sent.
Other rOpenSci community activities continue. 

**Please check back here again after 25 May when we will be announcing plans to slowly start back up.**

We express our continued great appreciation for the work of our authors and reviewers. Stay healthy and take care of one other.

The rOpenSci Editorial Board
⚠️⚠️⚠️⚠️⚠️

**Comment by geanders** [comment]
@RaoulWolf : I'm taking a look through to catch up on any package reviews that have stalled, and I see that we waiting to get an update from you regarding increased code coverage for tests in this package before we proceed. Our policy is to close package review issues one year after the author's last answer, so I wanted to check to see if you had any updates?

**Comment by maelle** [comment]
It's been more than one year after the author's last answer, so closing.

**Comment by RaoulWolf** [comment]
@maelle Sincerest apologies for the late reply - I haven't received any notifications regarding this submission until now! All answers since November 2019 are new to me... Oh my!

I am currently on vacation but I'm still very keen on pursuing `chemspiderapi` further. We have been using it routinely in our work, and last time I checked it provided more functionality than what was (is?) available over at `webchem`.

I haven't checked the progress with regard to possible API testing (as I haven't received notifications... again apologies!), but if you'd be willing to re-open this submission I'd be happy to do so.

**Comment by maelle** [comment]
Oh, in this case yep I'll re-open the issue!

**Comment by stitam** [comment]
Hi @RaoulWolf, I am a maintainer of `webchem`. Please note `webchem` is in heavy development, we have had three releases in the past 12 months including v1.0.0. The ChemSpider API has been fixed for about a year now and I think your package has significant overlap with `webchem`, only `webchem` provides access to other webservices as well. A paper using the fixed ChemSpider API in `webchem` has recently been published (https://www.jstatsoft.org/article/view/v093i13). Happy to discuss.

**Comment by maelle** [comment]
:wave: @RaoulWolf! Seeing that webchem wraps the ChemSpider API again I am a bit wary of [overlap](https://devguide.ropensci.org/policies.html#overlap). Is there anything webchem does not support?

**Comment by RaoulWolf** [comment]
@maelle thank you for re-opening the issue! Appreciated

@stitam very valid point. I'll get back to this in about a week from now. Thanks for the heads-up! (and for the record, I'm a `webchem` fan 😉)

**Comment by maelle** [comment]
:wave: @RaoulWolf

**Comment by RaoulWolf** [comment]
Lengthy post ahead, so watch out!

I took some time to compare the solutions of `webchem` and compare them to what is offered in `chemspiderapi`. The first step to adress the potential big overlap is to compare the available functionality, starting with everything ChemSpider offers (https://developer.rsc.org/docs/compounds-v1-trial/1/overview), and comparing the availability between `chemspiderapi` and `webchem` *(I tried to be as thorough as possible, but please forgive me if I oversaw functionalities within `webchem`!)*:

**FILTERING**

| ChemSpider compound API                  | `chemspiderapi` wrapper               | `webchem` wrapper|
| :--------------------------------------- | :------------------------------------ | :-----------------------: |
| filter-element-post                      | `post_element()`                      |            ?            |
| filter-formula-batch-post                | `post_formula_batch()`                |            ?            |
| filter-formula-batch-queryId-results-get | `get_formula_batch_queryId_results()` |            ?            |
| filter-formula-batch-queryId-status-get  | `get_formula_batch_queryId_status()`  |           ?            |
| filter-formula-post                      | `post_formula()`                      |            `cs_formula_csid()`            |
| filter-inchi-post                        | `post_inchi()`                        |            `cs_inchi_csid()`            |
| filter-inchikey-post                     | `post_inchikey()`                     |            `cs_inchikey_csid()`            |
| filter-intrinsicproperty-post            | `post_intrinsicproperty()`            |            ?            |
| filter-mass-batch-post                   | `post_mass_batch()`                   |            ?            |
| filter-mass-batch-queryId-results-get    | `get_mass_batch_queryId_results()`    |            ?           |
| filter-mass-batch-queryId-status-get     | `get_mass_batch_queryId_status()`     |            ?            |
| filter-mass-post                         | `post_mass()`                         |            ?           |
| filter-name-post                         | `post_name()`                         |           `cs_name_csid()`            |
| filter-queryId-results-get               | `get_queryId_results()`               |            `cs_query_csid()`            |
| filter-queryId-results-sdf-get           | `get_queryId_results_sdf()`           |            ?           |
| filter-queryId-status-get                | `get_queryId_status()`                |            ?            |
| filter-smiles-post                       | `post_smiles()`                       |            `cs_smiles_csid()`           |

**LOOKUPS**

| ChemSpider compound API | `chemspiderapi` wrapper | `webchem` wrapper |
| :---------------------- | :---------------------- | :-----------------------: |
| lookups-datasources-get | `get_datasources()`     |            `cs_datasources()`            |

**RECORDS**

| ChemSpider compound API                 | `chemspiderapi` wrapper             | `webchem` wrapper |
| :-------------------------------------- | :---------------------------------- | :-----------------------: |
| records-batch-post                      | `post_batch()`                      |            ?             |
| records-recordId-details-get            | `get_recordId_details()`            |            `cs_compinfo()`/`cs_extcompinfo()`            |
| records-recordId-externalreferences-get | `get_recordId_externalreferences()` |            ?             |
| records-recordId-image-get              | `get_recordId_image()`              |            `cs_img()`             |
| records-recordId-mol-get                | `get_recordId_mol()`                |            ?             |

**TOOLS**

| ChemSpider compound API      | `chemspiderapi` wrapper    | `webchem` wrapper |
| :--------------------------- | :------------------------- | :-----------------------: |
| tools-convert-post           | `post_convert()`           |           `cs_convert()`/`cs_convert_multiple()`             |
| tools-validate-inchikey-post | `post_validate_inchikey()` |            ?             |

While there (quite unsurprisingly) is overlap, `chemspiderapi` offers direct access to *all* API functionalities of ChemSpider. 

Another major difference between the solution offered by `chemspiderapi` and `webchem` is the workflow. It's really two different philosophies in my view, so there's no right or wrong. `chemspiderapi` is very explicit (and maybe lengthy) in every step, but "forces" users to familiarize themselves with the desired API use workflow of ChemSpider; additionally, `chemspiderapi` offers several vignettes to guide users in ways how to store their API token(s), rate-limiting quires, memoising queries, and saving .mol, .sdf, or .png files. `webchem` has an elegant "under the hood" approach to the ChemSpider's API functionalities it offers. 

Maybe less interesting, but also worth noting, is the unsurprising difference in dependencies. `chemspiderapi` has two dependencies, and `webchem` has 13 dependencies. `chemspiderapi` was designed to have as little dependencies as possible, and both dependencies (`curl` and `jsonlite`) are well-established and introduce no other dependencies. 

While for most users the functionalities within `webchem` are likely enough, I still think `chemspiderapi` provides a more "total" ChemSpider API experience, for anyone wishing to go down this rabbit hole. 

Let's discuss!

**Comment by maelle** [comment]
Thanks a lot for your detailed answer! Could you please add a column or sentences explaining to humans what e.g. _records-batch-post_ does i.e. the type of functionalities your package adds? I have the impression your package allows for posting information, what are the use cases for that? And what about the functionalities it adds that have nothing to do with posting?

Thanks again!

**Comment by RaoulWolf** [comment]
Right, apologies for not providing enough context! 

The different `post_*()` and `get_*()` functionalities are named so from the HTTP methods ([POST](https://en.wikipedia.org/wiki/POST_(HTTP)) and [GET](https://en.wikipedia.org/wiki/GET_(HTTP)), respectively).

In simple terms, all `post_*()` functions "upload" information into ChemSpider's API services (and usually get a query ID in return), and all `get_*()` functionalities "download" information from ChemSpider's API services.

In the example case you mentioned (*records-batch-post*), a list of up to 100 ChemSpider record IDs is POST-ed to ChemSpider's API services, and a query ID is returned. 

I hope this answers your question!

**Comment by maelle** [comment]
Thanks, and in what kinds of workflows do you need to post information?

@stitam could you please confirm the question marks in the comparison table, i.e. that webchem does not provide that functionality?

**Comment by RaoulWolf** [comment]
The principle workflows for the filtering functionalities all follow the same pattern: 

* POST query, 
* GET status, and 
* GET results. 

This is also mentioned in the [README](https://github.com/NIVANorge/chemspiderapi/blob/master/README.md) of `chemspiderapi`. The lookups, records and tools functionalities work directly, i.e., without the (manual) three step process.

**Comment by stitam** [comment]
Hi All! @RaoulWolf thanks for your detailed answer to @maelle's questions.

@RaoulWolf is right in saying that the two packages follow different philosophies and support different types or workflows. My understanding is that `chemspiderapi` aims to develop an R function for each ChemSpider API. On the other end, `webchem` focuses on the user experience, and aims to distance the user from the API, so instead of being "process" focused it is more "outcome" focused. I agree with @RaoulWolf in that there is no good or bad, these are two different approaches.

It follows from the difference in philosophies that `webchem` doesn't implement all APIs. The beginning of a `webchem` workflow is that a user is looking for a "specific" compound or set of compounds. Therefore we skipped a few APIs like `filter-mass-post` where the user would specify a molecular weight range and the API (actually 2-3 APIs called after eachother) would return a list of CSIDs for compounds with molecular weights falling in that range. At this point we didn't find a good enough use case for that functionality and our user haven't asked for it either.

We also didn't implement functions that process batch requests. We do see the benefit of batch requests, but not all requests have a batch alternative at the moment, so it was difficult to bind them all into a simple and easy to use user facing function. We did implement the non-batch alternatives for these queries however. ChemSpider APIs seem to be under development so once batch is available for all the queries we need (it's actualy `filter-name-batch-post` we are missing) then we'll add the batch options as well to make our queries more efficient.

Finally a few notes on the table above, in terms of user experience, in `webchem` all ChemSpider APIs that ultimately return CSIDs are actually pooled into `get_csid()`, that is the exported function. Also `get_queryId_status` is called within the functions to query the status before requesting the response itself. Other than these, yes @maelle, I can confirm the question marks, those APIs are currently not implemented in `webchem` so there is no overlap there at the moment.

**Comment by jooolia** [comment]
### Editor checks:

- [X] **Documentation**: The package has sufficient documentation available online (README, pkgdown docs) to allow for an assessment of functionality and scope without installing the package.
- [X] **Fit**: The package meets criteria for [fit](https://devguide.ropensci.org/policies.html#package-categories) and [overlap](https://devguide.ropensci.org/policies.html#overlap)
- [X] **Automated tests:** Package has a testing suite and is tested via Travis-CI or another CI service.
- [X] **License:** The package has a CRAN or OSI accepted license
- [X] **Repository:** The repository link resolves correctly
- [ ] **Archive** (JOSS only, may be post-review): The repository DOI resolves correctly
- [ ] **Version** (JOSS only, may be post-review): Does the release version given match the GitHub release (v1.0.0)?


Editor comments:

Hi @RaoulWolf 

Thank you for all of the comments and answers to our questions that you have provided. We will move ahead with the onboarding process for this package. I will be the editor handling the process going forward.

- We have been happy to see the discussion with the maintainers of {webchem} and think that it would be useful to mention webchem in your documentation. It is also a hope that these packages could work well together in the future (i.e. perhaps chemspiderapi could become a dependency for webchem or some other scenario).
 - We do [require 75 % code coverage for a package to go to review]((https://devguide.ropensci.org/building.html#testing) (and currently the code coverage is at 41% for this package), but I understand there is some difficulty with testing functions using an API. Some of the other editors have suggested [{presser}](https://r-lib.github.io/presser/) as a way to help create some of these tests.  

 Here are the results of some of the issues flagged with some of our initial editor checks:

```
── GP chemspiderapi 

It is good practice to

  ✖ write unit tests for all functions, and all package code in general. 41% of code lines are covered by test cases.

    R/FILTERING-get_formula_batch_queryId_results.R:29:NA
    R/FILTERING-get_formula_batch_queryId_results.R:31:NA
    R/FILTERING-get_formula_batch_queryId_results.R:33:NA
    R/FILTERING-get_formula_batch_queryId_results.R:35:NA
    R/FILTERING-get_formula_batch_queryId_results.R:37:NA
    ... and 414 more lines
```

- [ ] increase code coverage of unit tests
```

  ✖ add a "BugReports" field to DESCRIPTION, and point it to a bug tracker. Many online code hosting services provide
    bug trackers for free, https://github.com, https://gitlab.com, etc.

```

- [ ]  add a "BugReports" field to description

```
  ✖ avoid long code lines, it is bad for readability. Also, many people prefer editor windows that are about 80
    characters wide. Try make your lines shorter than 80 characters

    R/CHECKING-check_apikey.R:8:1
    R/CHECKING-check_apikey.R:12:1
    R/CHECKING-check_apikey.R:16:1
    R/CHECKING-check_complexity.R:8:1
    R/CHECKING-check_elements.R:12:1
    ... and 683 more lines

```
- [ ] check code line length (where possible)


I will look for package reviewers once these issues have been addressed. Please let me know if you have any questions or if I can clarify anything. 

Thanks! Julia

**Comment by RaoulWolf** [comment]
Hi @jooolia, thanks for the update! I'll try to fix the easier issues (bug reports and code line length) this week, and tackle code coverage next week. I haven't checked `presser` yet, thanks for the heads up! I'll let you know once the issues are addressed.

Thanks for the effort! Raoul

**Comment by RaoulWolf** [comment]
The `BugReports` field was added to the description, the `README` now mentions `{webchem}`, and I tried to minimize the line lengths as much as possible. 

All changes are live at https://github.com/NIVANorge/chemspiderapi

Code coverage hasn't been addressed yet.

**Comment by jooolia** [comment]
Thanks for keeping us posted on your updates @RaoulWolf!

**Comment by RaoulWolf** [comment]
A small update regarding test coverage. 

I tried implementing API tests with both {httptest} and {vcr}, but with no luck. {vcr} does not seem to support {curl} just yet (the [introduction](https://cran.r-project.org/web/packages/vcr/vignettes/vcr.html) hints at future support), and {httptest} simply runs the actual queries without mocking. 

I am now taking a look at {presser} and will let you know how far I get.

**Comment by jooolia** [comment]
Thanks for the update @RaoulWolf. If you get really stuck let us know.

**Comment by maelle** [comment]
{presser} supports curl indeed and that's the only HTTP testing package that does. :-)

**Comment by RaoulWolf** [comment]
Small update: I finally managed to wrap my head around {presser} and the first tests are running (see [here](https://github.com/NIVANorge/chemspiderapi/blob/master/tests/testthat/test-get_datasources.R), line 77). I'll now try to extend the testing functionality to the other functions. Baby steps 😃 

Meanwhile, I've also added GitHub Actions as CI (trying to substitute CircleCI and Travis CI). Unfortunately the {curl} requirement keeps breaking the installation on Linux machines. My initial approach was adding the following chunk to `R-CMD-check.yaml`:

````
     - name: Install libcurl
        if: runner.os == 'Linux'
        run: |
          sudo apt-get install -y libcurl4-openssl-dev
````

The test then fails when (trying to) installing {chemspiderapi}:
````
Error in dyn.load(file, DLLpath = DLLpath, ...) : 
  unable to load shared object '/home/runner/work/_temp/Library/curl/libs/curl.so':
  /usr/lib/x86_64-linux-gnu/libcurl.so.4: version `CURL_OPENSSL_3' not found (required by /home/runner/work/_temp/Library/curl/libs/curl.so)
````

I also tried with `libcurl4-gnutls-dev`, but no luck either. Any tips warmly welcome!

**Comment by maelle** [comment]
Untested tip, I'd look at what other curl dependencies have in their workflows e.g. https://github.com/r-lib/httr/blob/cb4e20c9e0b38c0c020a8756db8db7a882288eaf/.github/workflows/R-CMD-check.yaml#L60

**Comment by RaoulWolf** [comment]
Turns out the "trick" is to not use Ubuntu 20.xx at all, but good ol' 16.04. Now all tests pass 👍

**Comment by jooolia** [comment]
Thanks for keeping us updated @RaoulWolf. Have you run into any other roadblocks or is it looking like {presser} will help you with the testing?

**Comment by RaoulWolf** [comment]
Hi @jooolia, It first looked like {presser} would help and I can definitely setup mock-APIs with it (hurray!). But I still struggle to increase coverage. I suppose it's because the API request itself happens within {chemspiderapi} functions, but I'm not sure. An example can be seen [here](https://github.com/NIVANorge/chemspiderapi/blob/master/tests/testthat/test-get_datasources.R), line 77. The tests pass, but the coverage is not increased, which is slightly frustrating. On the positive side I have now over 500 tests for the package that pass 😅 

For the CI I was wondering if GitHub actions would be enough? I guess there's no need for Travis, AppVeyor or CircleCI?

**Comment by maelle** [comment]
Commenting on this again, hope it's fine. The presser package now has a function with which you don't need to do setup/teardown
* https://r-lib.github.io/presser/dev/reference/local_app_process.html
* https://r-lib.github.io/presser/dev/articles/how-to.html#how-do-i-start-the-app-when-writing-the-tests-

**Comment by jooolia** [comment]
Hi @RaoulWolf,
Yes the CI using GitHub actions should be sufficient as you have it implemented. 

I am also trying to wrap my head around webfakes  (previously presser now at https://r-lib.github.io/webfakes/ ) so I cannot help a lot with that aspect (but I will try as I am curious about how to do this sort of testing), however one thing that I see when I run `covr::report()` is that there are many arguments in the functions that are not tested in the tests so these lines are never run and thus not covered.  And this is also true with the API testing, since none of the functions from the package are called in the mock API the coverage does not increase. 

I will try to look a bit more at webfakes over the next few days and see if I can help.

@maelle your inputs and wisdom are always welcome. :)

**Comment by jooolia** [comment]
(Just putting this here for reference, there is a nice book by @maelle about http testing: https://books.ropensci.org/http-testing/packages-for-http-testing.html)

**Comment by maelle** [comment]
Thanks @jooolia :blue_heart: Demo of webfakes at https://github.com/ropensci-books/http-testing/pull/47 with demos of other packages.

So it seems the low coverage is not a webfakes problem, correct? I'm happy to help if needed.

**Comment by jooolia** [comment]
Wonderful @maelle! I think the demos will be very helpful! Thanks for pointing us to this material.

**Comment by RaoulWolf** [comment]
Hi @jooolia and @maelle, very cool to see {presser} become {webfakes}! I'll keep an eye out on the book 😃 

I'll try a few more variants to increase coverage with {webfakes} over the weekend; maybe there's a way I haven't thought about...

**Comment by RaoulWolf** [comment]
I tried now several other ways to code tests so it can run with "mock" APIs using {webfakes}, but I simply cannot find a way to emulate a (hardcoded URL) API from outside the function which defines the API itself. 

All other tests work well, but this seems very complicated to put in place at this point. I looked into {webchem}'s [approach](https://github.com/ropensci/webchem/blob/master/tests/testthat/test-chemspider.R) to this, and it seems they - apart from the very reasonable `skip_on_cran()` and `skip_on_ci()` controls - actually run real queries. 

Given my current subscription with ChemSpider I'd very much like to avoid running alot of queries every time the package is updated/revised.

How do we proceed now? I'm very much open to give {webfakes} another try with some help, but otherwise I seem to be stuck in a dead-end when it comes to increasing test coverage.

**Comment by maelle** [comment]
Where is that URL hard-coded? Could you use an internal function to define the API URL instead as in https://books.ropensci.org/http-testing/webfakes.html#actual-testing-2?

Reg real requests we've added this chapter to the book https://books.ropensci.org/http-testing/real-requests-chapter.html

**Comment by RaoulWolf** [comment]
The respective URLs are hard-coded in the respective functions, for example [here](https://github.com/NIVANorge/chemspiderapi/blob/master/R/FILTERING-post_inchikey.R): 
````
url <- "https://api.rsc.org/compounds/v1/filter/inchikey" 
````
This URL (and all others) are then used in `curl::curl_fetch_memory()`. Most of the API functionalities come with dedicated URLs, so this way of hard-coding allows to easier keep track of the URLs, in case there are any future changes. 

Additionally, it is sometimes necessary to modify the hard-coded URL depending on user-defined input, such as [here](https://github.com/NIVANorge/chemspiderapi/blob/master/R/FILTERING-get_queryId_results.R), where the minimum requirement is to provide a `queryId`, and `start` and `count` are optional: 
````
if (is.null(start) && is.null(count)) {
    url <- paste0("https://api.rsc.org/compounds/v1/filter/", queryId, "/results")
  }
  
  if (!is.null(start) && is.null(count)) {
    url <- paste0("https://api.rsc.org/compounds/v1/filter/", queryId, "/results?start=", start)
  }
  
  if (is.null(start) && !is.null(count)) {
    url <- paste0("https://api.rsc.org/compounds/v1/filter/", queryId, "/results?count=", count)
  }
  
  if (!is.null(start) && !is.null(count)) {
    url <- paste0("https://api.rsc.org/compounds/v1/filter/", queryId, "/results?start=", start, "&count=", count)
  }
````
I'll take a look into the "Making real requests" chapter of the book asap! This makes me wonder if there's a sensible way to run "real" API tests without exceeding rate-limitations; possibly a brute-force use of `Sys.sleep()` could help? Then again, if the tests are excluded from running on CI systems (`skip_on_ci()`), the testing coverage will not increase either...

**Comment by maelle** [comment]
Instead of having

```r
url <- "https://api.rsc.org/compounds/v1/filter/inchikey"
```

you could have

```r
url <- get_compounds_inchikey_url()
```

where `get_compounds_inchikey_url()` is

```r
get_compounds_inchikey_url <- function() {
  api_url <- Sys.getenv("COMPOUNDS_INCHIKEY_URL")

  if (nzchar(api_url)) {
    return(api_url)
  }

  "https://api.rsc.org/compounds/v1/filter/inchikey"
}
```

This way you can override the URL by setting the environment variable to the URL of a webfakes app process.

**Comment by maelle** [comment]
Same for your other snippet, you'd replace "https://api.rsc.org/compounds/v1/filter/" with `get_compounds_filter_url()`. Does it make sense?

**Comment by maelle** [comment]
> This makes me wonder if there's a sensible way to run "real" API tests without exceeding rate-limitations; possibly a brute-force use of Sys.sleep() could help? Then again, if the tests are excluded from running on CI systems (skip_on_ci()), the testing coverage will not increase either...

You'd make sure not to run them too often. Another thing that might help, or not, is asking the web API providers if you can get some sort of special developer account (I mention this in the book, for some APIs it can help as package maintainers are helping spread their product/API, after all).

True about test coverage, hopefully by making it possible to override the API URLs you can use more webfakes apps.

**Comment by maelle** [comment]
Also, if you don't mind my asking: why do you use curl by the way? (rather than httr or crul)

**Comment by RaoulWolf** [comment]
Lot's to unpack here, thank you for the detailed recommendations @maelle! Let's start from the end:

> Also, if you don't mind my asking: why do you use curl by the way? (rather than httr or crul)

when I first started working on the package, {curl} was the only package that would allow to run all different flavors of API queries that were necessary. For the record, in the beginning I tried implementing the functionalities with {httr} and {crul}, but not all API queries were possible. Apart from that {curl} comes with no dependencies, which I personally prefer (that's also the reason why I use {jsonlite} in the package). 

> Another thing that might help, or not, is asking the web API providers if you can get some sort of special developer account 

...let's just say we should explore other options first ;)

> This way you can override the URL by setting the environment variable to the URL of a webfakes app process.

I like the idea of using environmental variables, {webchem} uses it as well if I'm not mistaken. My worry is that the readability (and, after all, maintenance) of the code becomes more difficult with these constructs. To me this seems like a trade-off situation between code simpleness/tidiness on one hand and increasing test coverage on the other hand. But I think these thoughts are a bit bigger than my humble package :) I'll try to see if this approach works!

**Comment by maelle** [comment]
A more compact version would be

```r
baseurl <- function() {
  Sys.getenv("COMPOUNDS_INCHIKEY_URL", "https://api.rsc.org/compounds/v1/filter/inchikey")
}
```

Reg https://github.com/NIVANorge/chemspiderapi/blob/master/R/FILTERING-get_queryId_results.R?rgh-link-date=2020-12-08T14%3A32%3A34Z you might want to create a helper function that returns the URL depending on the variables (to have less if/else)?

**Comment by RaoulWolf** [comment]
I now implemented {webfakes} for most API calls and the test coverage is around 97%. CI is now fully switched to Github Actions. Everything is live at the repository: https://github.com/NIVANorge/chemspiderapi

Anything else needed?

**Comment by jooolia** [comment]
Great @RaoulWolf! I will have a closer look tomorrow evening. Thanks for the perseverance.

**Comment by jooolia** [comment]
Thanks for all the additions and updates @RaoulWolf. Indeed the test coverage now looks good. I will move forward with looking for reviewers. Thanks!

**Comment by jooolia** [comment]
Reviewer: @rajarshi
Reviewer: @yufree
Due date: 2021-03-15

Thanks for agreeing to review. The due date for your reviews is March 15th. Thanks! Julia

**Comment by yufree** [comment]
## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- **Briefly describe any working relationship you have (had) with the package authors.**
- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (If you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [x] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [ ] **Vignette(s)** demonstrating major functionality that runs successfully locally
    need extra work to fix some errors.
- [x] **Function Documentation:** for all exported functions
- [ ] **Examples** (that run successfully locally) for all exported functions
    need to fix some issues.
- [x] **Community guidelines** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

>##### For packages co-submitting to JOSS
>
>- [ ] The package has an **obvious research application** according to [JOSS's definition](https://joss.theoj.org/about#submission_requirements)
>
>The package contains a `paper.md` matching [JOSS's requirements](https://joss.theoj.org/about#paper_structure) with:
>
>- [ ] **A short summary** describing the high-level functionality of the software
>- [ ] **Authors:**  A list of authors with their affiliations
>- [ ] **A statement of need** clearly stating problems the software is designed to solve and its target audience.
>- [ ] **References:** with DOIs for all those that have one (e.g. papers, datasets, software).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software been confirmed.
- [x] **Performance:** Any performance claims of the software been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package
   and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [x] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines

Estimated hours spent reviewing: 7 hours

- [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

Overall it's a useful package for researchers to retrieve information from ChemSpider API. I tested this package as a chemical researcher and software developer. As a researcher, I might need the example of every function could return results without errors. As a software developer, I would check if the package have the proper style and readable code for other R users. At current stage, this package still need some improvements.

#### Information about Chemspider API

It would be helpful to mention that free account will have up to 1000 API requests per month in the readme. 

#### `goodpractice` and `spelling` check

For `goodpractice` check, lots of code lines were longer than 80 characters. I checked those lines and found some long code lines might need to format in a shorter way for the reader. Developer can consider `Reformat Code` function in RStudio to improve the codes. Another issues were about the test coverage, which looks fine for me.

For `spelling` check, most of the results were terms in Cheminformatics, which should be fine. However, in Rate-Limiting_API_Queries.Rmd:21, 'recomendations' should be 'recommendations'.

#### `inst` folder

I found the files in the `inst` folder and I wonder if the developer could mention this file somewhere for the user as a short introduction of this package. Or the developer could consider to move the contents to `<package-name>.R` and use `roxygen2` package to generate the Rd file for the whole package.

#### Function example

I think something might changed during the past year and some examples of function will not work. Because most of examples were protected by do-not-run with API token, regular test will not find those issues. I hope the developer could test those examples in the revision.

- The major issue in the example is that status is a list object and it's always needed to get the third element. In this case, `status=status` should be changed to `unlist(status)[3]` to get the 'Complete' status. Almost every function's example need to change this part to make sure the example could be run.

- Examples of `get_formula_batch_queryId_status`, `get_mass_batch_queryId_results`, `get_queryId_results_sdf` return error

- Example of `get_recordId_externalreferences` should return results instead of Null list for the user


#### vignette: Rate-Limiting API Queries

Line 91. "All queries need to report the status `Complete` before proceeding." and unlist(y)[1] only return one element as 1. The third element is the 'Complete' and that will give the chemspider ID. I wonder if this issue is from the change of API? Meanwhile, the returned object should be a vector of chemspider ID. This line could be changed to 

```r
demo_chemicals$id = mapply(FUN = function(x, y) sleepy_get_queryId_results(queryId = x, status = unlist(y)[3], apikey = apikey, coerce = T)$results, x = demo_chemicals$queryId, y = demo_chemicals$status)
```
Line 151. Similar issue as line 91

Line 189. Here the code should be 

```r
mutate(queryId = map_chr(InChIKey, ~ slowly_post_inchikey(inchikey = .x, apikey = apikey)$queryId))
```

Otherwise user will get the error "Error: Problem with `mutate()` input `queryId`.
x Can't coerce element 1 from a list to a character
ℹ Input `queryId` is `map_chr(InChIKey, ~slowly_post_inchikey(inchikey = .x, apikey = apikey))`."

Line 200. The code should be

```r
mutate(status = map(queryId, ~ unlist(slowly_get_queryId_status(queryId = .x, apikey = apikey)[3]))) %>% 
```

Line 212. The code should be

```
mutate(id = map2_int(queryId, status, ~ slowly_get_queryId_results(queryId = .x, status = .y, apikey = apikey)$results))
```

#### vignette: Remembering API Queries

This is an empty vignette. Maybe remove this till it's finished.

#### vignette: Saving MOL and SDF Files

Line 32-34. Maybe remove this chunk.

Line 44. Should be `writeLines(caffeine_mol$sdf, con = con)`

Line 54. Should be `queryId <- post_mass(mass = 194, range = 0.002, apikey = apikey)$queryId`

Line 66. Should be `caffeine_sdf <- get_queryId_results_sdf(queryId = queryId, status = status$status, apikey = apikey)`

#### vignette: Saving PNG Images

Actually `png` package could do this work. Maybe we don't need `magick` package unless we need extra features.

```r
caffeine_image_raw <- get_recordId_image(recordId = 2424L, apikey = apikey,decode = T)
png::writePNG(png::readPNG(caffeine_image_raw$image), "mypng.png")
```

In general, it's nice to see a package for ChemSpider API. As discussed above, `webchem` might have overlap functions while `chemspiderapi` give access to all of the APIs provided by ChemSpider. I hope those mentioned issues could be fixed.

Hope you will find the comments useful.

Thanks for the review invitation.

Miao

**Comment by RaoulWolf** [comment]
Thanks for the detailed review and comments @yufree!

I'll work through your comments and critiques during the next two weeks (busy times...) and will post the responses here. Once again thank you for the review!

**Comment by jooolia** [comment]
Thanks for the review @yufree.
Just a friendly reminder @rajarshi and @data-datum that the review due date is today. 
Thanks! Julia

**Comment by rajarshi** [comment]
## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- **Briefly describe any working relationship you have (had) with the package authors.**
- [X] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (If you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [X] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README
- [X] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [X] **Vignette(s)** demonstrating major functionality that runs successfully locally
- [X] **Function Documentation:** for all exported functions
- [X] **Examples** (that run successfully locally) for all exported functions
- [ ] **Community guidelines** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

>##### For packages co-submitting to JOSS
>
>- [ ] The package has an **obvious research application** according to [JOSS's definition](https://joss.theoj.org/about#submission_requirements)
>
>The package contains a `paper.md` matching [JOSS's requirements](https://joss.theoj.org/about#paper_structure) with:
>
>- [ ] **A short summary** describing the high-level functionality of the software
>- [ ] **Authors:**  A list of authors with their affiliations
>- [ ] **A statement of need** clearly stating problems the software is designed to solve and its target audience.
>- [ ] **References:** with DOIs for all those that have one (e.g. papers, datasets, software).

#### Functionality

- [X] **Installation:** Installation succeeds as documented.
- [X] **Functionality:** Any functional claims of the software been confirmed.
- [ ] **Performance:** Any performance claims of the software been confirmed.
- [ ] **Automated tests:** Unit tests cover essential functions of the package
   and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [X] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines

Estimated hours spent reviewing: 2

- [X] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

The package provide a consistent way to access Chemspider API endpoints. As already noted, there is overlap with `webchem`, and given the more expansive nature of this package, it may be worth considering merging this into `webchem` so that users can go to one place to access ChemSpider from R

It would be useful to note in the README  a link to the RSC dev pages where a user can get an API key

I'm not a fan of the combination of camelCase and underscore's in function names. For example `get_recordId_mol` could be `get_record_id_mol` or `getRecordIdMol` (with the former being preferred).

The [Remembering API Queries](https://github.com/NIVANorge/chemspiderapi/blob/master/vignettes/Remembering_API_Queries.Rmd) vignette seems to be empty

The rate limiting vignette is a great example of responsibly interacting with web APIs

Having to specify the api key with every call seems inefficient. One way to avoid this would be to have an `env` that is not exported, so that a user could set the key at the beginning of a session, which would add the key to the `env` and then functions would internally get the value of the key from the env. It would be good practice to have the `apikey` argument still present but with a default of `NULL`, so that if the user wanted to temporarily use a different key, they could.

Overall a well written package, that provides a comprehensive interface to the ChemSpider API

**Comment by RaoulWolf** [comment]
just a heads-up that the revisions are a bit delayed until after Easter

thank you again @yufree and @rajarshi for the reviews, greatly appreciated!

**Comment by data-datum** [comment]
## Package review

Please check off boxes as applicable, and elaborate in comments below. Your review is not limited to these topics, as described in the reviewer guide

- [X] Briefly describe any working relationship you have (had) with the package authors.

As the reviewer I confirm that there are no conflicts of interest for me to review this work (If you are unsure whether you are in conflict, please speak to your editor before starting your review).

**Documentation**

The package includes all the following forms of documentation:

- [X] A statement of need clearly stating problems the software is designed to solve and its target audience in README

- [X] Installation instructions: for the development version of package and any non-standard dependencies in README

- [ ] Vignette(s) demonstrating major functionality that runs successfully locally need extra work to fix some errors.

- [X] Function Documentation: for all exported functions

- [ ] Examples (that run successfully locally) for all exported functionsneed to fix some issues.

- [ ] Community guidelines including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with URL, BugReports and Maintainer (which may be autogenerated via Authors@R).

> For packages co-submitting to JOSS
> 
> The package has an obvious research application according to JOSS's definition
> 
> The package contains a paper.md matching JOSS's requirements with:
> 
> A short summary describing the high-level functionality of the software
> Authors: A list of authors with their affiliations
> A statement of need clearly stating problems the software is designed to solve and its target audience.
> References: with DOIs for all those that have one (e.g. papers, datasets, software).
> 


**Functionality**

- [X] Installation: Installation succeeds as documented.

- [X] Functionality: Any functional claims of the software been confirmed.

- [ ] Performance: Any performance claims of the software been confirmed.

- [ ] Automated tests: Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.

- [ X] Packaging guidelines: The package conforms to the rOpenSci packaging guidelines

Estimated hours spent reviewing: 4

- [X] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

### Review Comments 
* The “Remembering API Queries” vignette is empty. 
* The examples are not clear. It has a legend “Not Run”. 
* It would be great to add a CONTRIBUTING section.

**Comment by jooolia** [comment]
Thanks @yufree, @rajarshi  and @data-datum for the reviews and thank @RaoulWolf for the update on when we will expect revisions.

**Comment by jooolia** [comment]
Hi @RaoulWolf,
Have you had time to work on any of the comments from the reviewers? Thanks, Julia

**Comment by RaoulWolf** [comment]
Hi @jooolia,
still working on it! I had to postpone the revision process quite dramatically but hopefully by the end of June everything should be in order 👍

**Comment by RaoulWolf** [comment]
Hi again @jooolia , and sorry for the delay. I've started working through the comments of the reviewers and have addressed most of the issues already - I'll update this thread once the revision is complete. 

Meanwhile, the main repository for the package will be [RaoulWolf/chemspiderapi](https://github.com/RaoulWolf/chemspiderapi) after all (and not the previous repository); from August on I will not have access to the previous repository anymore and it will likely become private. Should the "new" repository ([RaoulWolf/chemspiderapi](https://github.com/RaoulWolf/chemspiderapi)) be a fork of the old one (like it is now) or rather a fresh repository?

**Comment by jooolia** [comment]
Hi @RaoulWolf ,
Thanks for keeping us updated on your progress.

I do not see any reason that the fork repository would be problematic. I will update the information at the top of this issue with this new repo. After the review process is completed we will transfer it to the rOpenSci organization. @maelle , do you think there would be any problems if we need to transfer a repository that is a fork (instead of the original)?

**Comment by jooolia** [comment]
Hi @RaoulWolf,
Do you have any updates on your revisions? If you are not able to work on it right now we could  put the "holding" label here. https://devguide.ropensci.org/policies.html?q=hold#policiesreviewprocess

Thanks! Julia

**Comment by jooolia** [comment]
Email sent to raoul.wolf@hotmail.de and raoul.wolf@niva.no asking for an update. Otherwise I will close this issue if we do not receive a response in a week( by 8 November 2021).

**Comment by jooolia** [comment]
Response to email received today:

> Hi Julia,
> 
> thanks for reaching out! I've been quite busy the past few months (new job...) but finished most of the requested changes already, except a revised vignette. I hope to be able to dedicate some more time to it before Christmas, so a "holding" label would be fantastic.
> 
> Once again, thank you for reaching out and have a great day!

Thus I will add the holding label until 2022 and then we can revisit.

**Comment by jooolia** [comment]
Email sent to raoul.wolf@hotmail.de

> Hi Raoul,
I am reaching out again regarding chemspiderapi. Do you have any updates? I think I will close the issue and if you have time to complete it you could reopen the issue and ping me on it.

>Thanks, Julia

With that I will close this issue. Cheers, Julia
